src/methods.jl

```diff
@@ -10,318 +19,446 @@ derivatives.
+            )
+        end
 
-julia> fdm(central_fdm(2, 1), exp, 0, Val(true))
-(FiniteDifferenceMethod:
-  order of method:       2
-  order of derivative:   1
-  grid:                  [-1, 1]
-  coefficients:          [-0.5, 0.5]
-  roundoff error:        1.42e-14
-  bounds on derivatives: 1.00e+02
-  step size:             1.69e-08
-  accuracy:              1.69e-06
-, 1.0000000031817473)
-```
-"""
-function fdm(
-    m::M,
-    f,
-    x::T,
-    ::Val{true};
-    condition=DEFAULT_CONDITION,
-    bound=_estimate_bound(f(x), condition),
-    eps=add_tiny(Base.eps(bound)),
-    adapt=m.history.adapt,
-    max_step=convert(T, 0.1),
-) where {T<:AbstractFloat, M<:FiniteDifferenceMethod}
-    if M <: Nonstandard && adapt > 0
-        throw(ArgumentError("can't adaptively compute bounds over Nonstandard grids"))
+        @doc """
+    $($(Meta.quot(fdm_fun)))(
+        p::Int,
+        q::Int;
+        adapt::Int=1,
+        condition::Int=DEFAULT_CONDITION,
+        geom::Bool=false
+    )
+
+Contruct a finite difference method at a $($(Meta.quot(direction))) grid of `p` linearly
+spaced points.
+
+# Arguments
+- `p::Int`: Number of grid points.
+- `q::Int`: Order of the derivative to estimate.
+
+# Keywords
+- `adapt::Int=1`: Use another finite difference method to estimate the magnitude of the
+    `p`th order derivative, which is important for the step size computation. Recurse
+    this procedure `adapt` times.
+- `condition::Int`: Condition number. See [`DEFAULT_CONDITION`](@ref).
+- `geom::Bool`: Use geometrically spaced points instead of linearly spaced points.
+
+# Returns
+- `FiniteDifferenceMethod`: The specified finite difference method.
+        """ $fdm_fun
     end
-    eps > 0 || throw(ArgumentError("eps must be positive, got $eps"))
-    bound > 0 || throw(ArgumentError("bound must be positive, got $bound"))
-    0 <= adapt < 20 - m.p || throw(ArgumentError("can't perform $adapt adaptation steps"))
-
-    # The below calculation can fail for `m.q == 0`, because then `C₂` may be zero. We
-    # therefore handle this edge case here.
-    m.q == 0 && return m, f(x)
+end
 
-    p = m.p
-    q = m.q
-    grid = m.grid
-    coefs = m.coefs
-
-    # Adaptively compute the bound on the function and derivative values, if applicable.
-    if adapt > 0
-        newm = (M.name.wrapper)(p + 1, p)
-        dfdx = fdm(
-            newm,
-            f,
-            x;
-            condition=condition,
-            eps=eps,
-            bound=bound,
-            max_step=max_step,
-            adapt=(adapt - 1),
+function _make_adaptive_bound_estimator(
+    constructor::Function,
+    q::Int,
+    adapt::Int,
+    condition::Int;
+    kw_args...
+)
+    if adapt >= 1
+        estimate_derivative = constructor(
+            q + 1, q, adapt=adapt - 1, condition=condition; kw_args...
         )
-        bound = _estimate_bound(dfdx, condition)
+        return (f, x) -> maximum(abs, estimate_derivative(f, x))
```

Comment:
`estimate_derivative` always returns a scalar so no need for `maximum` is there?
```suggestion
        return (f, x) -> abs(estimate_derivative(f, x))
```

Reply:
Ah, I forgot that the methods support functions `R -> R^n`, so it does need to be `maximum(abs, ...)`.